File: app01_java_block/src/main/java/com/gandarych/legalchain/contracts/AgriSupplyChainContract.java
package com.gandarych.legalchain.contracts;

import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Agriculture-sector contract: farm-to-fork supply chain transparency.
 *
 * <p>Each batch of produce accumulates an ordered, immutable trail of stage events
 * (harvest → processing → transport → warehouse → retail). The contract enforces that
 * stages only move <b>forward</b> — a shipment cannot "return" to the farm on paper,
 * which is precisely the class of document fraud this eliminates. Because every event
 * is a signed ledger transaction, any consumer, importer or food-safety authority
 * (e.g. under EU regulation 178/2002 traceability requirements) can audit the full
 * provenance of a batch without trusting any single participant's database.</p>
 */
@Component
public class AgriSupplyChainContract implements SmartContract {

    /** Ordered lifecycle stages; index = position in the lifecycle. */
    public static final List<String> STAGES =
            List.of("HARVEST", "PROCESSING", "TRANSPORT", "WAREHOUSE", "RETAIL");

    private final Map<String, List<SupplyChainEvent>> batches = new ConcurrentHashMap<>();

    @Override
    public String contractId() {
        return "agri-supply-chain-v1";
    }

    @Override
    public String description() {
        return "Farm-to-fork traceability: forward-only, signed stage events per batch "
                + "(harvest → processing → transport → warehouse → retail).";
    }

    @Override
    public ContractResult execute(Map<String, String> input) {
        String batchId = require(input, "batchId");
        String stage = require(input, "stage");
        String actor = require(input, "actor");
        String location = require(input, "location");
        String details = input.getOrDefault("details", "");

        int stageIndex = STAGES.indexOf(stage);
        if (stageIndex < 0) {
            throw new IllegalArgumentException("Unknown stage: " + stage + " (allowed: " + STAGES + ")");
        }

        List<SupplyChainEvent> trail = batches.computeIfAbsent(batchId, k -> new ArrayList<>());
        if (!trail.isEmpty()) {
            int lastIndex = STAGES.indexOf(trail.getLast().stage());
            if (stageIndex < lastIndex) {
                throw new IllegalArgumentException("Stage regression rejected: batch " + batchId
                        + " is already at " + trail.getLast().stage() + ", cannot record " + stage);
            }
        } else if (stageIndex != 0) {
            throw new IllegalArgumentException("Batch " + batchId + " must start at HARVEST");
        }

        SupplyChainEvent event = new SupplyChainEvent(
                batchId, stage, actor, location, details, System.currentTimeMillis());
        trail.add(event);

        return new ContractResult(
                contractId(), true,
                "Batch " + batchId + " recorded at stage " + stage + " (" + location + ")",
                Map.of("contract", contractId(), "batchId", batchId, "stage", stage,
                        "actor", actor, "location", location, "details", details),
                null);
    }

    /** Full provenance trail of a batch — the consumer/auditor view. */
    public List<SupplyChainEvent> trailOf(String batchId) {
        return List.copyOf(batches.getOrDefault(batchId, List.of()));
    }

    private static String require(Map<String, String> input, String key) {
        String value = input.get(key);
        if (value == null || value.isBlank()) {
            throw new IllegalArgumentException("Missing contract input: " + key);
        }
        return value;
    }

    public record SupplyChainEvent(String batchId, String stage, String actor,
                                   String location, String details, long timestamp) {
    }
}

File: app03_csharp_block/src/LegalChain/Contracts/AgriSupplyChainContract.cs
namespace LegalChain.Contracts;

/// <summary>
/// Agriculture-sector contract: farm-to-fork supply chain transparency.
/// </summary>
/// <remarks>
/// Each batch of produce accumulates an ordered, immutable trail of stage events
/// (harvest → processing → transport → warehouse → retail). The contract enforces that
/// stages only move <b>forward</b> — a shipment cannot "return" to the farm on paper,
/// which is precisely the class of document fraud this eliminates. Because every event
/// is a signed ledger transaction, any consumer, importer or food-safety authority
/// (e.g. under EU regulation 178/2002 traceability requirements) can audit the full
/// provenance of a batch without trusting any single participant's database.
/// </remarks>
public class AgriSupplyChainContract : ISmartContract
{
    /// <summary>Ordered lifecycle stages; index = position in the lifecycle.</summary>
    public static readonly IReadOnlyList<string> Stages =
        ["HARVEST", "PROCESSING", "TRANSPORT", "WAREHOUSE", "RETAIL"];

    private readonly Dictionary<string, List<SupplyChainEvent>> _batches = new();
    private readonly Lock _gate = new();

    public string ContractId => "agri-supply-chain-v1";

    public string Description
        => "Farm-to-fork traceability: forward-only, signed stage events per batch "
           + "(harvest → processing → transport → warehouse → retail).";

    public ContractResult Execute(IReadOnlyDictionary<string, string> input)
    {
        string batchId = Require(input, "batchId");
        string stage = Require(input, "stage");
        string actor = Require(input, "actor");
        string location = Require(input, "location");
        string details = input.GetValueOrDefault("details", "");

        int stageIndex = Stages.ToList().IndexOf(stage);
        if (stageIndex < 0)
        {
            throw new ArgumentException("Unknown stage: " + stage
                + " (allowed: [" + string.Join(", ", Stages) + "])");
        }

        SupplyChainEvent supplyEvent;
        lock (_gate)
        {
            if (!_batches.TryGetValue(batchId, out var trail))
            {
                _batches[batchId] = trail = new List<SupplyChainEvent>();
            }
            if (trail.Count > 0)
            {
                int lastIndex = Stages.ToList().IndexOf(trail[^1].Stage);
                if (stageIndex < lastIndex)
                {
                    throw new ArgumentException("Stage regression rejected: batch " + batchId
                        + " is already at " + trail[^1].Stage + ", cannot record " + stage);
                }
            }
            else if (stageIndex != 0)
            {
                throw new ArgumentException("Batch " + batchId + " must start at HARVEST");
            }

            supplyEvent = new SupplyChainEvent(
                batchId, stage, actor, location, details, DateTimeOffset.UtcNow.ToUnixTimeMilliseconds());
            trail.Add(supplyEvent);
        }

        return new ContractResult(
            ContractId, true,
            "Batch " + batchId + " recorded at stage " + stage + " (" + location + ")",
            new Dictionary<string, string>
            {
                ["contract"] = ContractId,
                ["batchId"] = batchId,
                ["stage"] = stage,
                ["actor"] = actor,
                ["location"] = location,
                ["details"] = details
            },
            null);
    }

    /// <summary>Full provenance trail of a batch — the consumer/auditor view.</summary>
    public IReadOnlyList<SupplyChainEvent> TrailOf(string batchId)
    {
        lock (_gate)
        {
            return _batches.TryGetValue(batchId, out var trail)
                ? trail.ToList()
                : [];
        }
    }

    private static string Require(IReadOnlyDictionary<string, string> input, string key)
    {
        if (!input.TryGetValue(key, out var value) || string.IsNullOrWhiteSpace(value))
        {
            throw new ArgumentException("Missing contract input: " + key);
        }
        return value;
    }

    public record SupplyChainEvent(string BatchId, string Stage, string Actor,
        string Location, string Details, long Timestamp);
}
